### ISSUE TITLE
Colorshake

### ISSUE THREAD

**Issue by AEFeinstein** [AEFeinstein/Swadge-Devkit-Fw#66]
The goal of Colorshake is to provide clear, causal LED fun. Every action should have a predictable reaction. The LEDs should be aesthetically pleasing and on more often than off while the Swadge is in motion.

Colorshake will have seven different sub-modes to translate motion into color. When the mode starts, the first sub-mode will automatically start.

The left button will cycle between the seven modes. The right button will adjust the brightness, cycling between three preset brightness. Brightness is adjusted by dividing each color channel before sending to the LEDs. The divisors should be 0x01, 0x08, and 0x40.

Each sub-mode may provide OLED graphics to help the user understand the correlation between action and reaction, and to also look cool. Showing the rotating ball on screen is the best example of this. The BPM waveform is a good start, but it needs labels for the numbers, so the user can understand what each is.

The seven sub-modes are as follows:
1. This mode uses BPM detection. A color wheel is displayed around the LEDs, the same as in `danceTimerMode4()`. When the BPM is stationary, the wheel does not move. When a BPM is detected, the wheel will spin at some multiple of the BPM. I'm guessing `(BPM / 128 == RPM)` will be appropriate, or something around there. Using 128 as the divisor makes division easy for the ESP8266.
1. This mode uses BPM detection. The BPM is mapped directly to a hue. I'm guessing `(BPM * 2 % 256)` will be appropriate. This hue can be passed to `EHSVtoHEX()`. The color should be displayed for enough time that the LEDs are never off while being shaken. If the Swadge is stationary for a while, the LEDs may be off.
1. This mode uses shock detection. The LEDs should start at off. When the magnitude of the acceleration vector crosses a threshold, all LEDs should be immediately set to maximum brightness and then slowly fade out over 1.5 seconds. The color of the LEDs should rotate around the color wheel so that each time the threshold is crossed, the hue is incremented. When the Swadge is shaken, the LEDs brightness will have a sawtooth pattern like `|\|\|\|\`.
Magnitude here is defined as `((x*x)+(y*y)+(z*z))`. There's no need to take the square root. The threshold is 400,000.
1. This mode uses shock detection. The LEDs should start at a dim red. When the magnitude of the acceleration vector crosses a threshold (see above), the LEDs will become chaotic (blinking random colors quickly, every 200ms or so). The LEDs will then settle back into a dim red (blink randomly at longer intervals with more red, until stable
1. This mode uses a single simulated ball. This is roll mode 6
1. This mode uses three balls for each color channel. This is roll mode 10 or 11, whichever is more aesthetically pleasing
1. This mode only uses orientation, no physics. Each axis (X, Y, Z) will be mapped one-to-one to a color channel (R, G, B). Any vector >255 should be clamped to 255.

**Comment by bbkiwi** [comment]
I think I can do something that behaves similar to what you describe, but not necessarily  in the way 
you suggest. 

Brightness settings are more complicated than simply using the three divisors. A mode where the the lights dim or change in intensity may look ok for the brightest setting, but simply scaling down won't look good,  either becomes jumpy, also colors mix of rgb often change hues. The use of gamma is attempt to fix this.

I've got 3 ways to estimate BPM but over the short term none are very accurate. The BPM tends to fluctuate wildly unless the movement is very rhythmical (ie not just shaking). 

The numbers currently at the top of the screen are BPM (method 1),  BPM (method 2), Total Energy
These are for my development use and are not meant to appear in the final version.

1. Total Energy (TE) would probably be better here and is a surogate for BPM in the sense that most people shake harder when they shake faster. So the hard the shake the faster the color wheel turns.
2. Again TE might work better. It will be easy to test BPM too.
3., 4. Shock detection will probably work better on the magnitude of high pass filtered input.  Note that joust using something like this by defining movement as deviation for a running average. 
3. If the shock induced fading color is to last 1.5 sec, should shocks within that period be ignored?
4. Similar comment
7. This really becomes tilt a color unless use movement relative to a average rest position.

**Comment by AEFeinstein** [comment]
I know linear-looking brightness is more complex than using a simple divisor, but it's a simple, good enough solution for these LEDs. Sometimes software engineering is about doing easy, good enough things, rather than perfectionist things. The divisor approach worked well with last year's Swadge, so it's worth sticking with.

1. I agree that Total Energy (TE) sounds like a better alternative to BPM. The important part is that shaking harder correlates to spinning faster, which TE accomplishes. I don't know what UI you want for this, but displaying the TE number with a label might be nice.
1. If you think TE looks better here, then go with that. The important part is that the LEDs correlate somewhat to shaking rhythm and stay pretty consistently on while the Swadge is being shaken.
1. My shock detection idea was naive and simple. If Joust is using deviation from a running average, that would be appropriate here too. Do you know how many milliseconds of input lag the high pass filter adds? I suspect it isn't much, but a fast response to shocks here is more important than Joust.
Any shock should reset the LEDs to maximum brightness. They should not be ignored furing the fading period. This way the LEDs will still blink to each shock, even if they don't fade out all the way.
1. See 3 for comments about shock detection.
1. No comments, ball is good.
1. No comments, balls are good.
1. Yes, it really is tilt-a-color. We thought it would be fine to have one mode which uses an absolute position rather than a relative position. This should be quite easy to implement.

**Comment by AEFeinstein** [comment]
@bbkiwi I thought a little more about mode 7 (tilt-a-color), and I think we want to map each axis to color without any discontinuities in brightness. This means that the mode should use the absolute value of the reported gravity vector, like such:

![image](https://user-images.githubusercontent.com/231180/69161624-71a3a900-0ab9-11ea-973f-75ff95a4ab71.png)

Reported values below -256 or above 256 should still be clamped to that range.

**Comment by AEFeinstein** [comment]
Notes from our earlier call here.

On the UI, it doesn't need a selector menu, it should just start at `beat_spin`. After tweaking and debugging is done, the numbers on the OLED should be replaced with the name of each mode. The name will be displayed for 10 seconds before being removed so the only thing seen is the waveform or ball or other graphic.

1. `beat_spin` / `power_shake` - I wrote `(BPM / 128 == RPM)` before but I meant `(BPM / 128 == RPM)`, so 60x faster. Oops, my mistake. Either way, the calculation is based on total energy now, and when giving it a good shake it should be about 128 RPM. Both the modes are similar, except that `beat_spin` has 6 LEDs and a smaller 'energy reserve' while `power_shake` has one LED and a larger 'energy reserve`. It might be worth meeting in the middle. Having three LEDs would make the spinning more noticeable and a medium 'energy reserve' would be pretty fun.
1. `beat_select` - Working well
1. `shock_change` - still working on it
1. `shock_chaotic` - still working on it
1. `roll_ball` - Working well
1. `roll_3_balls` - missing the green and blue balls
1. `tilt_a_color` - Working well

New modes
1. `twirl_a_color` - Meant to spin through the color wheel when the swadge is spun around. Fun, but we're not going to encourage swinging swadges on lanyards
1. `pov_effect` - The idea is to strobe RGB in succession quickly enough that it looks white when held still. When moved around, you'd see separate R G and B. Not accelerometer based, but motion sensitive anyway. As long as the strobing isn't seizure-inducing, it's pretty interesting.
1. `dft_shake` - Probably going to remove this

**Comment by bbkiwi** [comment]
1. No title window now. R button changes submode, name displays on OLED for 3 sec and slides away for 1 sec.

**Comment by AEFeinstein** [comment]
Slide? Nice. Animations beat sharp transitions any day.

**Comment by AEFeinstein** [comment]
Note to self, this is being built in https://github.com/bbkiwi/Swadge-Devkit-Fw/tree/colorshake

**Comment by AEFeinstein** [comment]
Notes to self after testing https://github.com/AEFeinstein/Swadge-Devkit-Fw/pull/72. ~~Generally, replace underscores in names with spaces.~~
1. ~~`beat_spin` - Great spinning at medium speed, could be faster when shaking vigorously (if the accelerometer can tell)~~
1. ~~`beat_select` - perfect~~
1. ~~`shock_change` - Love the graphics and effect. Threshold for shock detection could be a bit higher. It triggers when being picked up from a table now, which is a bit sensitive~~
1. ~~`shock_chaotic` - Same sensitivity comment as above. Should also reuse `shock_change` graphics.~~
1. ~~`roll_ball` - perfect~~
1. ~~`roll_3_balls` - perfect~~
1. ~~`tilt_a_color` - could use graphics, maybe some triangular representation of the three vector magnitudes?~~
1. ~~`twirl_a_color` - could use graphics, see above~~
1. ~~`pov_effect` - still looks strobey, not white, when at rest. I can try to strobe the LEDs faster, but if I can't this may be removed. Could also use graphics if we keep it~~
1. ~~`power_shake` - could use graphics, maybe a battery icon that fills up when shaken and depletes over time~~

**Comment by AEFeinstein** [comment]
For `pov_effect`, it's already strobing at `#define UPDATE_TIME_MS 16`, there's a note that it can't handle 8ms, and I'm cautious about changing timing for everything, so I'm going to remove it.

**Comment by bbkiwi** [comment]
Comment in text

> Generally, replace underscores in names with spaces.
Ok you could change display names to more funky ones.
> beat_spin - Great spinning at medium speed, could be faster when shaking vigorously (if the accelerometer can tell)
Two params control. step size and rev per beat
> beat_select - perfect
> shock_change - Love the graphics and effect. Threshold for shock detection could be a bit higher. It triggers when being picked up from a table now, which is a bit sensitive
Biff, boom, whap 
> shock_chaotic - Same sensitivity comment as above. Should also reuse shock_change graphics.
I thought no graphics was better as to not distract  from the leds which are more interesting than change.
> 1, roll_ball - perfect
> roll_3_balls - perfect
> tilt_a_color - could use graphics, maybe some triangular representation of the three vector magnitudes?
I'm easy here, wee sprite or dot in the triangle. However certain colors never happen eg white, the color space explored in this case is 2 dimentional.
> twirl_a_color - could use graphics, see above
I prefer not just in case you twirl it with the cord:-) Also the color space is just 1 dimensional eg angle.
> pov_effect - still looks strobey, not white, when at rest. I can try to strobe the LEDs faster, but if I can't this may be removed. Could also use graphics if we keep it
Now it cycles over 4 frames @ 60 fps ... Color, off, complement  of color, off... You could try just flpping color, complement to double the rate. In the main display there is code that prints the max time spent. It could be modified to reset when each sub mode starts. It was the roll ball sub modes that take close to 8 ms. The others are much faster so maybe the rate could be upped fir POV. 
> power_shake - could use graphics, maybe a battery icon that fills up when shaken and depletes over time
Good idea. May be a banana that fills up!

**Comment by AEFeinstein** [comment]
As far as I could tell `beat_spin` would spin at a fixed speed as long as `(smoothActivity > ACTIVITY_BOUND)`, so I changed it to cue off `bpmFromCrossing` instead. This way there's a linear relation between shaking it faster and the wheel spinning faster.

I like the principle of no graphics to not distract from LEDs, but that only works if all modes have no graphics. As soon as some modes have graphics, it becomes a requirement for all of them. For `tilt_a_color` I get that the space explored is 2D, but there are three separate color vectors that can be used to adjust the vertices of a triangle. I wrote it up and think it looks good. It's less good for `twirl_a_color` where it's a 1D space. I'll have to think of something else.
